### PR TITLE
DEP: deprecate exec_command

### LINKED ISSUE #11980: MAINT: exec_command() cleanup
With the impending demise of `exec_command()` related to #11824, there will soon only be unit tests that call this function in our code base even though it will no longer be used.

The unit tests are in `numpy/distutils/tests/test_exec_command.py` -- will we want to remove the tests that specifically call `exec_command()`?  I assume we won't want to unit test the standard library stuff we've replaced `exec_command()` with.

Also, do we need a deprecation marker of some sort on `exec_command()` before removing it? I'm not sure how public-facing / used it is; maybe I need to check that deprecation policy NEP again!

### PR DESCRIPTION
**Pull request by mattip**
Deprectate `exec_command`, `temp_file_name`,  remove  `_supports_fileno`. None of these are used by numpy internally, only `exec_command` is tested. Fixes #11980. This is a conservative approach, we could also remove these without deprecation.

### CODE REVIEW

File: numpy/distutils/tests/test_exec_command.py
Comment (eric-wieser) [inline numpy/distutils/tests/test_exec_command.py]: These `with`s are getting out of hand!

**Review by eric-wieser — APPROVED** [review]
Needs a release note, but otherwise looks great to me.

**Comment by mattip** [comment]
this PR has one approval, can it be merged?

File: doc/release/1.17.0-notes.rst
Comment (rgommers) [inline doc/release/1.17.0-notes.rst]: These need double backticks (or they'll turn into broken links after we remove the functions).

And `dist_utils` is spelled `distutils`

File: numpy/distutils/tests/test_exec_command.py
Comment (rgommers) [reply numpy/distutils/tests/test_exec_command.py]: 4level `with`s is pretty bad/funny indeed:)

nothing to change, so resolving the comment

**Comment by rgommers** [comment]
LGTM, can be merged after the simple fix to the release notes I think

File: doc/release/1.17.0-notes.rst
Comment (mattip) [reply doc/release/1.17.0-notes.rst]: fixed

**Comment by rgommers** [comment]
Merged, thanks @mattip

**Pull request merged**
